### ISSUE TITLE
Saving game

### ISSUE THREAD

**Issue by axx0** [axx0/Civ2-clone#104]
So we can either:
1) save the game as in the original binary .sav format (this shouldn't be a problem as the whole .sav structure is more or less figured out)
2) save the game in a newer format which would make editing the saved game easier and new stuff could also be added, but this means you can't open the file with the original game (is this even important?)

**Comment by reubene** [comment]
Option 1 is a no go if we want more than 7 civs,

I think option 2 is the only option. I suggest a ziped json or bson format

**Comment by axx0** [comment]
Should the new sav file have a new extension (e.g. .savx) to distinguish it from the classic files?

**Comment by reubene** [comment]
It's not really necessary, you can tell what a file is from the contents

**Comment by axx0** [comment]
Go ahead, I'm busy with some other stuff.

**Comment by axx0** [comment]
I've done some changes to the way classic sav files are loaded - got rid of some files and made simplifications so that the stuff is more in line with what you're doing with json saves. We can also probably get rid of GameData down the line. Warn me if you notice something is not working properly.

Regarding your comment on TOT ruleset loading - the main folder should be Original (we can probably ignore the root TOT game folder, it mostly has a bunch of MGE files which don't serve any purpose, at least I think I came to that conclusion when testing it some time ago). I haven't tested any TOT stuff lately, if it's not behaving correctly I'll try to roll back to your changes.

**Comment by sirloreal** [comment]
Nice work, though I haven't really had a time to look through the code in too much detail for that area yet. If you have worries or concerns about the changes breaking something you could add some targetted Unit Tests next  to make sure your simplifications and any further additions to saving and loading still do what you expect. It's unlikely to be bullet-proof but it'll give you more assurance :)

**Comment by sirloreal** [comment]
I've made a startin my forked repository to get the LoadFrom() routine under test, against your code before the change. I'll build on it to add more validation for the classic save file and a json type file as the LoadFrom() method handles both. Idea is run the tests against old code, then run against your latest code a few weeks ago and if nothing broke, we good :) 
Then you'll have some coverage for further improvements to the saving mechanism.

**Comment by axx0** [comment]
Nice, will be interesting to see what you'll come up with.

**Comment by crhy** [comment]
Great work guys!  I love the idea of .savx for the new format, and having compatibility with the old .sav for old games.

**Comment by crhy** [comment]
When can we have a 0.1 release and get more eyes on the project? /me ducks.

**Comment by sirloreal** [comment]
Life hasn't had the plans I wanted over the winter (long story!) but I had some time today to get some tests in place, and they work with your before and after refactors. 

There's more detail in the PR but a couple of things I found while testing and playing around with the code, so I don't forget
- LoadFrom doesn't completely throw a FileNotFoundException as needed, so I'll fix that when I get a few minutes.
- The sav file I used came from the original game; my starting settler had a goody hut next to him (lucky me!) but it was missing when I loaded the sav into the code for this repo. I'll look into why that is in case there's a weakness or bug in the code to load.

**Comment by reubene** [comment]
I've looked over the PR it's a good start on tests, ultimately if this is to become widely used we will need a comprehensive test suite. 
  
> * The sav file I used came from the original game; my starting settler had a goody hut next to him (lucky me!) but it was missing when I loaded the sav into the code for this repo. I'll look into why that is in case there's a weakness or bug in the code to load.

I this this is because Goody huts are among the things not implemented at all in this version, they shouldn't be too hard to do though.  I think there is  formula to determine their locations like for special resources, and as I recall the game uses the barbarians visibility bit to indicate that they have been explored.

**Comment by axx0** [comment]
Works well on my side, good job.
I've discovered more stuff in sav files which needs to be added but I'm putting it off and I'll do a big change some time in the future. Until then expect goody huts and some other things to differ from the original.

**Comment by sirloreal** [comment]
Thanks for the feedback. I'll make a separate issue for adding Goody Huts.

**Comment by reubene** [comment]
Do you think you could make available the save files you're using for the tests? Perhaps check them in if they're not to big. That way other people can also run the same tests.

**Comment by sirloreal** [comment]
Yes absolutely I can. Sorry for the slow reply, I've been sick for the last week. I'll work on it this week and finish a bit of file handling refactoring I've been doing in my fork of the repo.
